Improve TAP tests

jeltz opened the pull request:
Having worked a bit too much with out TAP tests during the last week I realized that one way they could be improved while still using this way of prepending we are like `pg_regress` is to do it more properly instead of halfhearted so I added a `PGTDE::psql()` helper which calls runs the query and then appends both stderr and stdout to the file when later will diff.

The reason for this is because right now when something goes wrong and some query fails you just get a truncated output file and then have to start modifying the test case to add debug output. Ideally we should have the error message of the failing query easily visible when the test run fails.

What do you think? The alternative would to abandon the idea of diffing the two files and instead aim for more idiomatic TAP tests instead because just running `safe_psql()` does not seem to give us much.

Edit: This is ready for review now.

jeltz committed:
Convert output of TAP tests to be more like pg_regress

This improves readbility of diffs a lot.

jeltz committed:
Clean up TAP test helpers

Checking for PostgreSQL versions can be done in a cleaner way plus we do
not need to run chmod.

AndersAstrand commented on contrib/pg_tde/t/pgtde.pm:
```diff
@@ -56,20 +56,31 @@ sub pgtde_init_pg
 
-    # For Server version 15 & above, use PostgreSQL::Test::Utils to write to files
-    if ($PG_MAJOR_VERSION > 14) {
+    if ($PG_MAJOR_VERSION >= 15) {
```
I suspect this is a vestigial from when pg_tde lived in a separate repo, we don't really need to do this at all now since we know for sure what version of postgres we're testing.

otoh it might be nice to keep this for a possible future where we might be able to move pg_tde out of the main repo again. The question is how much of this we should do in our code in general.

jeltz replied:
Yeah, I was going back and forth on that.

AndersAstrand commented on contrib/pg_tde/t/pgtde.pm:
```diff
@@ -56,20 +56,31 @@ sub pgtde_init_pg
+{
+    my ($node, $dbname, $sql) = @_;
+
+    my (undef, $stdout, $stderr) = $node->psql($dbname, $sql, extra_params => ['-a', '-P', 'format=aligned', '-P', 'tuples_only=off']);
```
I like that we now would get a readable diff, if we had one :)

AndersAstrand reviewed (commented):
This looks like a great cleanup to me! However I'm not sure I follow the reasoning behind treating TAP tests as pg_regress tests, but without a usable diff. Knowing what's wrong when they fail is much trickier than it would have to be if we just used regular TAP expectations.

dAdAbird commented:
Looks good to me. It definitely brings much cleaner tests code and it's good to have diff on errors

jeltz committed:
Remove unused includes in TAP tests

Additionally we remove unnecessary copies if the generated config file.

jeltz committed:
Remove pointless white space and comments from TAP tests

It removed readability rather than improved it.

jeltz committed:
Always restart the server in the same way in our TAP tests

It is hard to keep track when every place does it differently.

Naeem-Akhter reviewed (commented):
It looks much better and is the way to go if we keep the diff-based option intact. 

Regarding whether we should opt for a diff-based approach or idiomatic TAP, I would prefer to have support for both, as we will always have the option to choose depending upon the use case (in case a diff is needed). However, I believe the idiomatic TAP test case should be our preferred approach moving forward.

Naeem-Akhter commented on contrib/pg_tde/t/pgtde.pm:
```diff
@@ -56,6 +56,21 @@ sub pgtde_init_pg
+    my ($node, $dbname, $sql) = @_;
+
+    my (undef, $stdout, $stderr) = $node->psql($dbname, $sql, extra_params => ['-a']);
+
```
IIRC, though, there is no support for function overloading in Perl. It would be nice to have a similar function where we just return the command return code or maybe a full ($ret, $stdout, $stderr).

jeltz replied:
Maybe, but until I have a user for it I do not think we should be build it.

Naeem-Akhter commented on contrib/pg_tde/t/pgtde.pm:
```diff
@@ -56,6 +56,21 @@ sub pgtde_init_pg
+}
+
 sub append_to_file
 {
```
Suggestion, Wouldn't changing it to `append_to_result_file `be better?

jeltz replied:
Yeah, that is a better name.

dutow commented on contrib/pg_tde/t/expected/001_basic.out:
```diff
@@ -1,16 +1,43 @@
+(1 row)
+
+CREATE TABLE test_enc(id SERIAL,k INTEGER,PRIMARY KEY (id)) USING tde_heap;
+psql:<stdin>:1: ERROR:  principal key not configured
```
It would be nice to also show the error codes here, that would give us one more reminder to fix them.

jeltz replied:
Maybe, the issue is that it is not in our SQL tests where I imagine most of these errors will be. Hmmm.

dutow reviewed (commented):
Looks much better!

jeltz committed:
Add test helper to TAP tests which runs psql and appends

This simplifies working with tests a lot since now we will also always
get the errors from failed queries directly in the test output instead
of it being truncated when the TAP tests aborts due to the query
failing.

There is still a good case for why we should instead write idiomatic TAP
tests but this at least does a lot to improve the expereince of people
who have to work with these tests without changing the way the tests
work. Plus that the code is cleaner now so it should be easier to move
away from this way of testing in the future.

jeltz committed:
Rename test helper append_to_file() to append_to_result_file()

This way we can also import PostgreSQL's test utilities into the
current namespace.

jeltz committed:
Make global variable local in TAP tests

jeltz commented:
@Naeem-Akhter I did not covert `contrib/pg_tde/t/005_multiple_extensions.pl` since I could not run it with a reasonable amout of effort. Plus I am not sure it belongs in this repo in the first place.

Naeem-Akhter commented:
> @Naeem-Akhter I did not covert `contrib/pg_tde/t/005_multiple_extensions.pl` since I could not run it with a reasonable amout of effort. Plus I am not sure it belongs in this repo in the first place.

To make this test case work, we need to add a GH workflow/action where we install all PSP/PPG-supported extensions using packages. This test case won't work and will be skipped against a source build, as we are not building any extension (pg_stat_monitor, plaudit, set_user, pg_repack PostGIS) as part of make process. 

So, please leave this test case as it is. I will take care of it when I add the GH workflow for testing against package installation.

AndersAstrand reviewed (approved):
This looks like a great cleanup!
